internal/util/logging/embeddable_handler.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"context"
+	"log/slog"
+	"maps"
+	"runtime"
+	"slices"
+	"strconv"
+	"sync"
+)
+
+// Embeddable handler is a [slog.Handler] for custom logger.
+// It is used to make sure that the config.LogLevel is effective.
+//
+//nolint:vet // for readability
+type embeddableHandler struct {
+	opts *NewHandlerOpts
+	ga   []groupOrAttrs
+
+	// m is used to protect opts.Handler from race conditions.
+	m *sync.Mutex
+
+	testAttrs map[string]any
+}
+
+// newEmbeddableHandler creates a new embeddable handler.
+func newEmbeddableHandler(opts *NewHandlerOpts, attrs map[string]any) *embeddableHandler {
+	return &embeddableHandler{opts: opts, testAttrs: attrs, m: new(sync.Mutex)}
+}
+
+// Enabled implements [slog.Handler].
+func (h *embeddableHandler) Enabled(ctx context.Context, l slog.Level) bool {
+	minLevel := slog.LevelInfo
+	if h.opts.Level != nil {
+		minLevel = h.opts.Level.Level()
+	}
+
+	return l >= minLevel
+}
+
+// Handle implements [slog.Handler].
+func (h *embeddableHandler) Handle(ctx context.Context, record slog.Record) error {
+	if h.testAttrs != nil {
```

Comment:
`h.testAttrs` is only used for tests. This means that this condition applies only for tests. Instead, we should make sure that options apply even if `h.testAttrs == nil` for normal cases. See implementation for other handlers :hugs:

Reply:
I think we also need to ignore the RemoveLevel/Source fields, cause we can not control the logic in  `embeddableHandler.Handle()` function by the user :)